File: src/Binary/DesignInMemoryFileSystem.java
package Binary;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// https://leetcode.ca/2017-07-10-588-Design-In-Memory-File-System/
class DesignInMemoryFileSystem {
    class Node {
        String path;
        boolean isFile;
        List<Node> children;

        public Node() {
            this("", false);
        }

        public Node(String path, boolean isFile) {
            this.path = path;
            this.isFile = isFile;
            if (!isFile)
                children = new ArrayList<>();
        }
    }

    Node root;
    Map<String, String> fileContentMap;

    public DesignInMemoryFileSystem() {
        root = new Node();
        fileContentMap = new HashMap<>();
    }

    public List<String> ls(String path) {
        String[] array = path.substring(1).split("/");
        Node node = root;
        int length = path.length() == 1 ? 0 : array.length;
        for (int i = 0; i < length; i++) {
            String subPath = array[i];
            int index = binarySearch(node.children, subPath);
            if (index >= 0)
                node = node.children.get(index);
            else
                return new ArrayList<>();
        }
        List<String> list = new ArrayList<>();
        if (node.isFile) {
            list.add(node.path);
        } else {
            List<Node> children = node.children;
            for (Node child : children) {
                list.add(child.path);
            }
        }
        return list;
    }

    public void mkdir(String path) {
        if (path.equals("/"))
            return;
        String[] array = path.substring(1).split("/");
        Node node = root;
        int length = array.length;
        for (int i = 0; i < length; i++) {
            String subPath = array[i];
            int index = binarySearch(node.children, subPath);
            if (index >= 0)
                node = node.children.get(index);
            else {
                Node newNode = new Node(subPath, false);
                node.children.add(-index - 1, newNode);
                node = newNode;
            }
        }
    }

    public void addContentToFile(String filePath, String content) {
        String[] array = filePath.substring(1).split("/");
        Node node = root;
        int length = array.length - 1;
        for (int i = 0; i < length; i++) {
            String subPath = array[i];
            int index = binarySearch(node.children, subPath);
            if (index >= 0)
                node = node.children.get(index);
            else {
                Node newNode = new Node(subPath, false);
                node.children.add(-index - 1, newNode);
                node = newNode;
            }
        }
        Node fileNode = new Node(array[length], true);
        int index = binarySearch(node.children, array[length]);
        if (index < 0)
            node.children.add(-index - 1, fileNode);
        String prevContent = fileContentMap.getOrDefault(filePath, "");
        String newContent = prevContent + content;
        fileContentMap.put(filePath, newContent);
    }

    public String readContentFromFile(String filePath) {
        return fileContentMap.getOrDefault(filePath, "");
    }

    private int binarySearch(List<Node> list, String key) {
        int low = 0, high = list.size() - 1;
        while (low <= high) {
            int mid = (high - low) / 2 + low;
            Node node = list.get(mid);
            if (node.path.equals(key))
                return mid;
            else if (node.path.compareTo(key) > 0)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return -low - 1;
    }
}

File: py/essential/DesignIn-MemoryFileSystem.py
# https://leetcode.ca/2017-07-10-588-Design-In-Memory-File-System/
# Grind

from typing import List


class Trie:
    def __init__(self):
        self.name = None
        self.isFile = False
        self.content = []
        self.children = {}

    def insert(self, path: str, isFile: bool):
        node = self
        ps = path.split('/')
        for p in ps[1:]:
            if p not in node.children:
                node.children[p] = Trie()
            node = node.children[p]
        node.isFile = isFile
        if isFile:
            node.name = ps[-1] # the file name
        return node

    def search(self, path: str):
        node = self
        if path == '/':
            return node
        ps = path.split('/')
        for p in ps[1:]:
            if p not in node.children:
                return None
            node = node.children[p]
        return node

class FileSystem:
    def __init__(self):
        self.root = Trie()

    def ls(self, path: str) -> List[str]:
        node = self.root.search(path)
        if node is None:
            return []
        if node.isFile:
            return [node.name]
        return sorted(node.children.keys())

    def mkdir(self, path: str) -> None:
        self.root.insert(path, False)

    def addContentToFile(self, filePath: str, content: str) -> None:
        node = self.root.insert(filePath, True)
        node.content.append(content)

    def readContentFromFile(self, filePath: str) -> str:
        node = self.root.search(filePath)
        return ''.join(node.content)
    
s=FileSystem()
s.mkdir("/a/b/c")
s.addContentToFile("/a/b/c/d.txt", "hello")
print(s.ls("/a/b/c")) # ["d.txt"]
print(s.readContentFromFile("/a/b/c/d.txt")) # "hello"
s.mkdir("/a/b/e")
print(s.ls("/a/b")) # ["c", "e"]
print(s.ls("/a")) # ["b"]
print(s.ls("/")) # ["a"]
s.addContentToFile("/a/b/c/d.txt", "hello1")
print(s.readContentFromFile("/a/b/c/d.txt")) # "hellohello1"